embassy-usb-hid/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+
+        Self {
+            input: ReportWriter { ep_in },
+            output: ReportReader {
+                ep_out,
+                receiver: &state.out_signal,
+            },
+        }
+    }
+
+    /// Gets the [`ReportWriter`] for input reports.
+    ///
+    /// **Note:** If the `HidClass` was created with [`new_ep_out()`](Self::new_ep_out)
+    /// this writer will be useless as no endpoint is availabe to send reports.
+    pub fn input(&mut self) -> &mut ReportWriter<'d, D, IN_N> {
+        &mut self.input
+    }
+
+    /// Gets the [`ReportReader`] for output reports.
+    pub fn output(&mut self) -> &mut ReportReader<'d, D, OUT_N> {
+        &mut self.output
+    }
+
+    /// Splits this `HidClass` into seperate readers/writers for input and output reports.
+    pub fn split(self) -> (ReportWriter<'d, D, IN_N>, ReportReader<'d, D, OUT_N>) {
+        (self.input, self.output)
+    }
+}
+
+pub struct ReportWriter<'d, D: Driver<'d>, const N: usize> {
+    ep_in: Option<D::EndpointIn>,
+}
+
+pub struct ReportReader<'d, D: Driver<'d>, const N: usize> {
+    ep_out: Option<D::EndpointOut>,
+    receiver: &'d Signal<(usize, [u8; N])>,
+}
+
+impl<'d, D: Driver<'d>, const N: usize> ReportWriter<'d, D, N> {
+    /// Tries to write an input report by serializing the given report structure.
+    ///
+    /// Panics if no endpoint is available.
+    #[cfg(feature = "usbd-hid")]
+    pub async fn serialize<IR: AsInputReport>(&mut self, r: &IR) -> Result<(), WriteError> {
+        let mut buf: [u8; N] = [0; N];
+        let size = match serialize(&mut buf, r) {
+            Ok(size) => size,
+            Err(_) => return Err(WriteError::BufferOverflow),
+        };
+        self.write(&buf[0..size]).await
+    }
+
+    /// Writes `report` to its interrupt endpoint.
+    ///
+    /// Panics if no endpoint is available.
+    pub async fn write(&mut self, report: &[u8]) -> Result<(), WriteError> {
+        assert!(report.len() <= N);
+
+        let ep = self
+            .ep_in
+            .as_mut()
+            .expect("An IN endpoint must be allocated to write input reports.");
+
+        let max_packet_size = usize::from(ep.info().max_packet_size);
+        let zlp_needed = report.len() < N && (report.len() % max_packet_size == 0);
+        for chunk in report.chunks(max_packet_size) {
+            ep.write(chunk).await?;
+        }
+
+        if zlp_needed {
+            ep.write(&[]).await?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<'d, D: Driver<'d>, const N: usize> ReportReader<'d, D, N> {
+    pub async fn read(&mut self, buf: &mut [u8]) -> Result<usize, ReadError> {
+        assert!(buf.len() >= N);
+        if let Some(ep) = &mut self.ep_out {
+            let max_packet_size = usize::from(ep.info().max_packet_size);
+
+            let mut chunks = buf.chunks_mut(max_packet_size);
+
+            // Wait until we've received a chunk from the endpoint or a report from a SET_REPORT control request
```

Comment:
The logic here is quite complex. Is it actually needed?

The [HID specs](https://www.usb.org/document-library/device-class-definition-hid-111), page 10 say:

> If a device declares an Interrupt Out endpoint then Output reports are transmitted by the host to the device through the Interrupt Out endpoint. If no Interrupt Out endpoint is declared then Output reports are transmitted to a device through the Control endpoint, using Set_Report(Output) requests.

I guess this means that if we *do* have the INTERRUPT OUT pipe then reports MUST go through it, and we can ignore the control pipe. This'd simplify the code a lot. Hopefully it holds true in practice...

Reply:
I think that should be true of well behaved hosts, but I don't think the spec is quite that strict and I'm not confident all hosts will operate that way. Page 55 of the spec seems looser, for example:

> Reports may also be requested (polled) and sent through the Control pipe or sent through an optional Interrupt Out pipe.

I'd rather err on the side of making the device work consistently regardless of how the host behaves.

Reply:
Hmm... `usbd-hid` outright rejects all control requests and it seems to work for people... :joy: 

Another problem with handling reports from the control pipe is it needs an extra `[u8; OUT_N]` of RAM for the signal, if we could get rid of that it'd be nice...

TinyUSB seems to [handle](https://github.com/hathach/tinyusb/blob/9c8c5c1c53214b8954bb28c33a496b5b423d5ecc/src/class/hid/hid_device.c#L326) [both](https://github.com/hathach/tinyusb/blob/9c8c5c1c53214b8954bb28c33a496b5b423d5ecc/src/class/hid/hid_device.c#L410) though, but maybe it's just to support the case of not having the OUT endpoint?

Not sure what's best... Perhaps we can check whether major OSs behave properly, then do the simplification if it is the case?

Reply:
It would be awfully nice if we could count on that. My fear is that some obscure host or an application using libusb directly or something would cause a device to suddenly break. The simplest solution is to always have SET_REPORT control requests route to `RequestHandler::set_report()` and OUT endpoint reports route to `HidClass::output()`. But I expect that will mean device developers who allocate an OUT endpoint won't handle SET_REPORT requests for output reports in `RequestHandler::set_report()` and if there are bad hosts those devices will break in the field. I want to think about it some more to see if I can come up with a low to zero overhead way of merging the streams.

Reply:
I did some more reading, it seems you're right, and we can't rely on reports always coming on the interrupt pipe :( Linux has APIs to explicitly send stuff on the control endpoint, for example.

I've commented in the main thread an alternative that I think could work though :)